### ISSUE TITLE
Handle array values for JSON columns

### ISSUE THREAD

**Issue by iamgergo** [laravel/framework#23684]
- Laravel Version: 5.6.12
- PHP Version: 7.2.3
- Database Driver & Version: MySQL & 5.7.21

### Description:

Bulk updating JSON columns with array values are not working properly. It throws an 
``Illuminate\Database\QueryException``, because it tries to convert ``array`` to ``string``:

Illuminate\Database\QueryException with message 'PHP Notice:  Array to string conversion in .../vendor/laravel/framework/src/Illuminate/Database/MySqlConnection.php on line 80 (SQL: update `posts` set `meta` = json_set(`meta`, '$."tags"', 1), `updated_at` = 2018-03-24 14:39:29 limit 10)'

### Steps To Reproduce:

I have a ``Post`` model. In the migration, I add the JSON column to the schema.

```php
$table->json('meta')
```
Inside the meta column, I want to store the tags as an array.

Since I don't want to update all the ``meta`` every the time, only the tags, I add the path to the fillable array on the model.

```php
$fillable = ['meta->tags']
```
After I saved a model and retrieve the meta attributes it should look like this.
```
$post->meta

[
    ...
    tags => [1, 2, 3, 4, 5],
    ...
]
```
If I grab a single model and try to update it, it works.

```php
$post->update(['meta->tags', [1,2,3]])
```
It updates only the ``tags`` attribute on the JSON column, and nothing else as I wanted.

Now if I try the same, but with multiple models, it fails.

```php
Post::take(10)->update(['meta->tags' => [1,2,3]])
```
After running this, it throws the error I inserted in the description.

I think it would be nice if we could update JSON attributes like normal attributes since the number of queries could be reduced.

**Comment by iamgergo** [comment]
We can work around this problem by modifying two files:

1. **Illuminate/Database/MySqlConnection.php**
2. **Illuminate/Database/Query/JsonExpression.php**

We could use ``json_array(?)`` in the ``JsonExpression`` class:

```php
case 'array':
    return 'json_array(?)';
```
But, since the ``json_array`` mysql function accepts standalone values and not a whole array, we would have to pass every value in the array one-by-one (and for the bindings as well). That means:
``json_array(1, 2, 3, 4)`` and not ``json_array([1, 2, 3, 4]).

So our parameter binding would look like this: ``json_array(?, ?, ?, ?)``. In the ``MySqlConnection``, where the value binding happens, we would handle these bindings as 4 unique binding, but still, accept one value which is an ``array``. So we have to stick with one binding.

We can be tricky and use ``json_merge`` here, which requires a string version of the ``array`` and a second one which could be empty.

```php
case 'array':
    return 'json_merge(?, "[]")';
```

Then one parameter binding is enough and all we have to do is to check if the value is an array and convert it to json. In the MySqlConnection:

```php
foreach ($bindings as $key => $value) {
    $statement->bindValue(
         is_string($key) ? $key : $key + 1,
         is_array($value) ? json_encode($value) : $value,
         is_int($value) || is_float($value) ? PDO::PARAM_INT : PDO::PARAM_STR
     );
}
```

This works perfectly for multiple and single updates as well.

**Comment by iamgergo** [comment]
This issue is solved in PR #23685 .
